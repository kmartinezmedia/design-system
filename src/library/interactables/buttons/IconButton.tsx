import React, { FunctionComponent } from 'react';
import {
  TouchableWithoutFeedback,
  GestureResponderEvent,
  Animated,
  StyleSheet,
} from 'react-native';
import { View, ViewProps } from '@designSystem/primitives';
import { usePressAnimation } from '@designSystem/hooks';
import { sizing } from '@designSystem/theme';

export interface IconButtonProps extends Omit<ViewProps, 'onPress'> {
  disabled?: boolean;
  onPress?: (event: GestureResponderEvent) => void;
  onLongPress?: (event: GestureResponderEvent) => void;
  testID?: string;
}

export const IconButton: FunctionComponent<IconButtonProps> = ({
  children,
  disabled,
  onPress,
  onLongPress,
  testID,
  ...props
}) => {
  const [pressIn, pressOut, scale] = usePressAnimation();
  return (
    <View {...props}>
      <TouchableWithoutFeedback
        onPress={onPress}
        onPressIn={pressIn}
        onPressOut={pressOut}
        onLongPress={onLongPress}
        disabled={disabled}>
        <Animated.View
          style={[
            styles.iconButton,
            disabled ? styles.disabled : undefined,
            { transform: [{ scale }] },
          ]}
          testID={testID}>
          {children}
        </Animated.View>
      </TouchableWithoutFeedback>
    </View>
  );
};

const styles = StyleSheet.create({
  iconButton: {
    alignItems: 'center',
    flexDirection: 'column',
    justifyContent: 'center',
    // todo: this should be more strict, but for backwards
    // compat until new icons land we're doing min*sizes
    minHeight: sizing.icons.button,
    minWidth: sizing.icons.button,
  },
  disabled: {
    opacity: 0.5,
  },
});
