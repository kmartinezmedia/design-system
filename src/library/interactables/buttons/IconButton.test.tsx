import React from 'react';
import { Text } from 'react-native';
import { create, act, ReactTestRenderer } from 'react-test-renderer';
import { IconButton } from './IconButton';

describe('IconButton', () => {
  it('default', () => {
    let component: ReactTestRenderer;
    act(() => {
      component = create(
        <>
          <IconButton>
            <Text>🐛</Text>
          </IconButton>
          <IconButton disabled>
            <Text>🐛</Text>
          </IconButton>
        </>,
      );
    });
    expect(component!.toJSON()).toMatchSnapshot();
  });
  it('positions', () => {
    let component: ReactTestRenderer;
    act(() => {
      component = create(
        <>
          <IconButton spacing={3}>
            <Text>🐛</Text>
          </IconButton>
          <IconButton spacing={2} pin="right">
            <Text>🐛</Text>
          </IconButton>
        </>,
      );
    });
    expect(component!.toJSON()).toMatchSnapshot();
  });
});
