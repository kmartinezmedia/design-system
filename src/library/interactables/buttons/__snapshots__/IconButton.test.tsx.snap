// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton default 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "alignSelf": "auto",
          "flexDirection": "column",
          "flexGrow": undefined,
          "flexShrink": undefined,
          "flexWrap": "nowrap",
          "height": undefined,
          "justifyContent": "flex-start",
          "width": undefined,
        },
        undefined,
        undefined,
        undefined,
        Object {},
      ]
    }
  >
    <TouchableWithoutFeedback
      onPressIn={[Function]}
      onPressOut={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "column",
            "justifyContent": "center",
            "minHeight": 40,
            "minWidth": 40,
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
          }
        }
      >
        <Text>
          🐛
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "alignSelf": "auto",
          "flexDirection": "column",
          "flexGrow": undefined,
          "flexShrink": undefined,
          "flexWrap": "nowrap",
          "height": undefined,
          "justifyContent": "flex-start",
          "width": undefined,
        },
        undefined,
        undefined,
        undefined,
        Object {},
      ]
    }
  >
    <TouchableWithoutFeedback
      disabled={true}
      onPressIn={[Function]}
      onPressOut={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "column",
            "justifyContent": "center",
            "minHeight": 40,
            "minWidth": 40,
            "opacity": 0.5,
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
          }
        }
      >
        <Text>
          🐛
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>,
]
`;

exports[`IconButton positions 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "alignSelf": "auto",
          "flexDirection": "column",
          "flexGrow": undefined,
          "flexShrink": undefined,
          "flexWrap": "nowrap",
          "height": undefined,
          "justifyContent": "flex-start",
          "width": undefined,
        },
        Object {
          "padding": 12,
        },
        undefined,
        undefined,
        Object {},
      ]
    }
  >
    <TouchableWithoutFeedback
      onPressIn={[Function]}
      onPressOut={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "column",
            "justifyContent": "center",
            "minHeight": 40,
            "minWidth": 40,
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
          }
        }
      >
        <Text>
          🐛
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>,
  <View
    style={
      Array [
        Object {
          "alignItems": "stretch",
          "alignSelf": "auto",
          "flexDirection": "column",
          "flexGrow": undefined,
          "flexShrink": undefined,
          "flexWrap": "nowrap",
          "height": undefined,
          "justifyContent": "flex-start",
          "width": undefined,
        },
        Object {
          "padding": 8,
        },
        Object {
          "bottom": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        undefined,
        Object {},
      ]
    }
  >
    <TouchableWithoutFeedback
      onPressIn={[Function]}
      onPressOut={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "column",
            "justifyContent": "center",
            "minHeight": 40,
            "minWidth": 40,
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
          }
        }
      >
        <Text>
          🐛
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>,
]
`;
